File: aoc/2022/day11/rust/src/main.rs
use core::panic;
use regex::{Captures, Regex};
use std::fs;

#[derive(Debug, Clone)]
struct Monkey {
    name: u64,
    test_against: u64,
    throw_to: (u64, u64),
    op: (String, String),
    inspected_count: u64,
    worry_levels: Vec<u64>,
}

fn solve(mut monkeys: Vec<Monkey>, is_part2: bool) -> u64 {
    let limit = if is_part2 { 10000 } else { 20 };

    let modulus: u64 = if is_part2 {
        monkeys.iter().map(|m| m.test_against).product()
    } else {
        0
    };

    for x in 1..=limit {
        for i in 0..monkeys.len() {
            let monkey = (monkeys[i]).to_owned();
            // println!("Monkey: {:?}", monkey);

            for value in monkey.worry_levels {
                let operand: u64;

                if let Ok(acc) = monkey.op.1.parse() {
                    operand = acc;
                } else {
                    operand = value.clone();
                };

                let mut worry = if monkey.op.0 == "*" {
                    value * operand
                } else if monkey.op.0 == "+" {
                    value + operand
                } else {
                    panic!("Unknown operator")
                };

                if is_part2 {
                    worry = worry % modulus;
                } else {
                    worry = worry / 3;
                }

                // println!("Worry: {worry}");

                if worry.clone() % monkey.test_against == 0 {
                    // println!("Throwing to: {}", monkey.throw_to.0);
                    monkeys[monkey.throw_to.0 as usize].worry_levels.push(worry);
                    monkeys[monkey.throw_to.0 as usize].inspected_count += 1;
                } else {
                    // println!("Throwing to: {}", monkey.throw_to.1);
                    monkeys[monkey.throw_to.1 as usize].worry_levels.push(worry);
                    monkeys[monkey.throw_to.1 as usize].inspected_count += 1;
                }

                monkeys[i].worry_levels.pop();
            }
        }

        // The items monkeys currently have in their hands & not inspected yet
        if x == limit {
            monkeys.iter_mut().for_each(|monkey| {
                monkey.inspected_count -= monkey.worry_levels.len() as u64;
            })
        }
    }

    let inspected_counts = monkeys
        .iter()
        .map(|m| m.inspected_count)
        .collect::<Vec<u64>>();

    // println!("Inspected counts: {:?}", inspected_counts);

    let (max, next_max) = largest_two(inspected_counts);

    max * next_max
}

fn main() {
    let Ok(monkeys) = get_parsed_data("../input.txt") else { return };

    println!("part_01: {:?}", solve(monkeys.clone(), false));
    println!("part_02: {:?}", solve(monkeys.clone(), true));
}

fn get_parsed_data(file_path: &str) -> Result<Vec<Monkey>, ()> {
    let Ok(data) = fs::read_to_string(file_path) else { return Err(()) };

    let mut monkeys = vec![];

    data.lines().enumerate().step_by(7).for_each(|(line, _)| {
        let mut monkey = Monkey {
            name: 0,
            test_against: 1,
            throw_to: (0, 0),
            inspected_count: 0,
            worry_levels: vec![],
            op: (String::new(), String::new()),
        };

        for l in line..=line + 6 {
            if l == line + 6 {
                monkeys.push(monkey.clone());
                continue;
            };

            Regex::new(r"\d+|\+|\*|(\d+|old)$")
                .expect("Couldn't parse regex")
                .captures_iter(
                    data.lines()
                        .nth(l)
                        .expect(&format!("Couldn't extract line {}", l)),
                )
                .collect::<Vec<Captures>>()
                .iter()
                .for_each(|cap: &Captures| {
                    let parsed = cap.get(0).expect("Empty capture group").as_str();

                    match l % 7 {
                        0 => {
                            monkey.name = parsed.parse::<u64>().expect("Couldn't parse monkey name")
                        }
                        1 => {
                            monkey
                                .worry_levels
                                .push(parsed.parse::<u64>().expect("Couldn't parse worry level"));
                            monkey.inspected_count += 1;
                        }
                        2 => {
                            if monkey.op.0 == String::new() {
                                monkey.op.0 = parsed.to_owned();
                            } else {
                                monkey.op.1 = parsed.to_owned();
                            }
                        }
                        3 => {
                            monkey.test_against =
                                parsed.parse::<u64>().expect("Couldn't parse divisor")
                        }
                        4 => {
                            monkey.throw_to.0 = parsed
                                .parse::<u64>()
                                .expect("Couldn't parse monkey for true case")
                        }
                        5 => {
                            monkey.throw_to.1 = parsed
                                .parse::<u64>()
                                .expect("Couldn't parse monkey for false case")
                        }
                        _ => panic!(),
                    };
                });
        }
    });

    Ok(monkeys)
}

fn largest_two(values: Vec<u64>) -> (u64, u64) {
    let (mut largest, mut second_largest) = (0, 0);

    for value in values {
        if value > largest {
            second_largest = largest;
            largest = value;
        } else if value > second_largest {
            second_largest = value;
        }
    }

    (largest, second_largest)
}

File: aoc/2022/day11/go/main.go
package main

import (
	"fmt"
	"math/big"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Monkey struct {
	Name           string
	TestAgainst    uint64
	ThrowTo        [2]uint64
	InspectedCount uint64
	WorryLevels    []big.Int
	Operations     [2]string
}

// Not working correctly; use rust code for correct implementation
func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic("Error reading file")
	}

	var r = regexp.MustCompile(`\d+|\+|\*|(\d+|old)$`)
	var lines = strings.SplitN(string(data), "\n", -1)
	var monkeys = make([]Monkey, 0)
	var limit int = 10000

	for i := 0; i < len(lines)/7; i++ {
		name := r.FindAllString(lines[i*7+0], -1)
		worryLevels := r.FindAllString(lines[i*7+1], -1)
		operations := r.FindAllString(lines[i*7+2], -1)
		testAgainst := r.FindAllString(lines[i*7+3], -1)
		caseTrue := r.FindAllString(lines[i*7+4], -1)
		caseFalse := r.FindAllString(lines[i*7+5], -1)

		monkeys = append(monkeys, Monkey{
			Name:           name[0],
			TestAgainst:    toUint32(testAgainst[0]),
			InspectedCount: uint64(len(worryLevels)),
			WorryLevels:    stringArrayToBigIntArray(worryLevels),
			Operations:     [2]string{operations[0], operations[1]},
			ThrowTo:        [2]uint64{toUint32(caseTrue[0]), toUint32(caseFalse[0])},
		})
	}

	for i := 1; i <= limit; i++ {
		for i, monkey := range monkeys {

			for _, w := range monkey.WorryLevels {
				var val big.Int

				v, err := strconv.Atoi(monkey.Operations[1])
				if err != nil {
					val = w
				} else {
					val = *big.NewInt(int64(v))
				}

				var worry big.Int

				switch monkey.Operations[0] {
				case "*":
					worry.Set(val.Mul(&val, &w))
				case "+":
					worry.Set(val.Add(&val, &w))
				default:
					panic("Unknown operator")
				}

				// worry.Div(&worry, big.NewInt(3))

				var worryCopy big.Int
				worryCopy.Set(&worry)

				// fmt.Println("Worry:", worry.Int64())

				if worryCopy.Rem(&worryCopy, big.NewInt(int64(monkey.TestAgainst))).Cmp(big.NewInt(0)) == 0 {
					monkeys[monkey.ThrowTo[0]].WorryLevels = append(monkeys[monkey.ThrowTo[0]].WorryLevels, worry)
					monkeys[monkey.ThrowTo[0]].InspectedCount += 1
					// println("Throwing to monkey", monkey.ThrowTo[0])
				} else {
					monkeys[monkey.ThrowTo[1]].WorryLevels = append(monkeys[monkey.ThrowTo[1]].WorryLevels, worry)
					monkeys[monkey.ThrowTo[1]].InspectedCount += 1
					// println("Throwing to monkey", monkey.ThrowTo[1])
				}

				monkeys[i].WorryLevels = monkeys[i].WorryLevels[1:]

			}
		}

		if i == limit /* || i%100 == 0 || i%20 == 0 */ {
			for m, monkey := range monkeys {
				monkeys[m].InspectedCount -= uint64(len(monkey.WorryLevels))
				// fmt.Println("Monkey", monkey.Name, "has", monkey.InspectedCount, "inspected values at", i)
			}
		}
	}

	var inspectedCounts []uint64

	for _, v := range monkeys {
		inspectedCounts = append(inspectedCounts, v.InspectedCount)
	}

	fmt.Println("Inspected counts:", inspectedCounts)
	fmt.Println("part_02:", productOfMaxTwo(inspectedCounts))
}

func productOfMaxTwo[T uint64](values []T) T {
	var max T = 0
	var secondMax T = 0

	for _, value := range values {
		if value > max {
			secondMax = max
			max = value
		} else if value > secondMax {
			secondMax = value
		}
	}

	println("Max two:", max, secondMax)
	return max * secondMax
}

func toUint32(value string) uint64 {
	num, err := strconv.Atoi(value)
	if err != nil {
		panic("Couldn't parse number")
	}
	return uint64(num)
}

func stringArrayToBigIntArray(arr []string) []big.Int {
	var returnArray = []big.Int{}
	for _, v := range arr {
		returnArray = append(returnArray, *big.NewInt(int64(toUint32(v))))
	}
	return returnArray
}
